# 以电影简介为例比较jieba和snownlp的处理准确度和处理时间

In [33]:
import pandas as pd
import jieba
import pickle
import csv
import re
from snownlp import SnowNLP
import time

In [37]:
def intro_process(text):
    global stop_words
    global alltag
    text = re.sub(r'[^\w\s]', ' ', text)
    # 分割句子
    sentences = re.split(r'[.!?]', text)
    # 去除多余的空白
    sentences = [s.strip() for s in sentences if s.strip()]
    if len(sentences)!=0:
        sentences=sentences[0].split()
    originaltag=[]
    origin=[]
    for sentence in sentences:
        originaltag=jieba.lcut(sentence)
        origin.append(originaltag)
    if len(origin)>0:
        origin=origin[0]
    kong=[' ']
    afterdelte=[]
    for word in origin:
        if word not in stop_words and word not in kong and word in alltag:
            afterdelte.append(word)
    return afterdelte
def snow_process(text):
    global stop_words
    global alltag
    text = re.sub(r'[^\w\s]', ' ', text)
    # 分割句子
    sentences = re.split(r'[.!?]', text)
    # 去除多余的空白
    sentences = [s.strip() for s in sentences if s.strip()]
    if len(sentences)!=0:
        sentences=sentences[0].split()
    originaltag=[]
    origin=[]
    for sentence in sentences:
        originaltag=SnowNLP(sentence).words
        origin.append(originaltag)
    if len(origin)>0:
        origin=origin[0]
    kong=[' ']
    afterdelte=[]
    for word in origin:
        if word not in stop_words and word not in kong and word in alltag:
            afterdelte.append(word)
    return afterdelte 
  
stop_list='../data/stopword.txt'
stop_words = []
with open(stop_list, 'r', encoding='utf-8') as stoplst:
    for line in stoplst.readlines():
        stop_words.append(line.strip())
neededinfo=[]
alltag=[]
filepath='../data/movie.csv'
with open(filepath, 'r',encoding='utf-8') as file:
    csv_reader = csv.reader(file)
    for row in csv_reader:
        neededinfo.append(row)
for tag in neededinfo:
    tags=tag_process(tag[5])
    for ta in tags:
        if ta not in alltag:
            alltag.append(ta)
    

## jieba从简介所得到的tag总数和花费时间

In [38]:
jiebatag=[]
start=time.time()
for data in neededinfo:
    jiebatag+=intro_process(data[13])
end=time.time()
print(len(jiebatag))
print(end-start)

206
1.6928455829620361


## snownlp从电影简介所得到的tag总数和花费时间

In [39]:
snowtag=[]
start=time.time()
for data in neededinfo:
    snowtag+=snow_process(data[13])
end=time.time()
print(len(snowtag))
print(end-start)

238
64.91031289100647


## 总结
jieba花费时间1.69s,得到tag数206
snownlp花费时间64.91s,得到tag数238
可见jieba花费时间少,精度较低,snownlp花费时间远大于jieba,精度更高